### PR TITLE
feat: Support specifying the pattern for transient and retryable errors

### PR DESCRIPTION
**Pull request by terrytangyuan**
Signed-off-by: terrytangyuan <terrytangyuan@gmail.com>

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -36,6 +36,10 @@ type Config struct {
 	// MainContainer holds container customization for the main container
 	MainContainer *apiv1.Container `json:"mainContainer,omitempty"`
 
+	// TransientErrorPattern specifies the pattern to match for errors that can be seen as transient
+	// and retryable.
+	TransientErrorPattern string `json:"transientErrorPattern,omitempty"`
```
Comment (alexec) [inline config/config.go:41]: I think this should be a rarely used feature, so an environment variable would be the way to configure it - it is much simpler too.

Comment (terrytangyuan) [reply config/config.go:41]: In our case, we'd like to whitelist non-critical errors that can be retried for issues during cluster maintenance/migration, etc., especially for non-production clusters where certain services may not be available.

Good suggestion. I'll add an env var to configure this instead.

Comment (terrytangyuan) [reply config/config.go]: Just updated. Please take another look.

File: util/errors/errors.go [hunk 2]
```diff
@@ -15,7 +17,18 @@ func IsTransientErr(err error) bool {
 		return false
 	}
 	err = argoerrs.Cause(err)
-	return isExceededQuotaErr(err) || apierr.IsTooManyRequests(err) || isResourceQuotaConflictErr(err) || isTransientNetworkErr(err) || apierr.IsServerTimeout(err) || apierr.IsServiceUnavailable(err)
+	// TRANSIENT_ERROR_PATTERN allows to specify the pattern to match for errors that can be seen as transient
+	// and retryable.
+	pattern, _ := os.LookupEnv("TRANSIENT_ERROR_PATTERN")
```
[the remark this answers is not in the document]
Comment (terrytangyuan) [reply util/errors/errors.go:26]: Thanks. Updated.

**Review by alexec — APPROVED** [review]

**Pull request merged**